### PR TITLE
Implement combiner tensors

### PR DESCRIPTION
**Pull request by kshyatt**
Some of this code could be cleaned up but it appears to work.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -8,11 +8,11 @@ using ITensors, Test
         "test_indexset.jl",
         "test_itensor.jl",
         "test_contract.jl",
+        "test_combiner.jl",
         "test_trg.jl",
         "test_ctmrg.jl",
         "test_dmrg.jl",
         "test_siteset.jl",
-        "test_phys_sitesets.jl",
```
Comment (emstoudenmire) [inline test/runtests.jl]: Did you intend to remove this line?

Comment (kshyatt) [inline test/runtests.jl]: what the

File: src/storage/combiner.jl [hunk 2]
```diff
@@ -0,0 +1,89 @@
+export Combiner
+
+struct CombinerStorage <: TensorStorage
+    ci::Index
+    inds::IndexSet
```
Comment (emstoudenmire) [inline src/storage/combiner.jl]: Did the code end up using `ci` and `inds` stored inside CombinerStorage? If not, perhaps CombinerStorage can just hold no data.

File: test/runtests.jl (hunk 1 above)
Comment (kshyatt) [reply test/runtests.jl]: no I don't know why it vanished :(. I'll put it back

File: src/storage/combiner.jl (hunk 2 above)
Comment (kshyatt) [reply src/storage/combiner.jl]: It doesn't yet but it might be nice to have the Combiner carry around which index is the combined one

Comment (emstoudenmire) [reply src/storage/combiner.jl]: True, but let’s make sure we have a concrete application of this choice. I think a good one would be a function `combinedIndex` which acts on an ITensor with combiner storage, and returns `ci`. We have that function in C++ but the way it works relies on the first index being the combined one, which is a brittle way to do it. 

So how about we keep `ci` but remove `inds`. We can always add `inds` back in later when it is needed without breaking any existing code.

Comment (kshyatt) [reply src/storage/combiner.jl]: Sure!

**Comment by kshyatt** [comment]
Better now?

**Comment by emstoudenmire** [comment]
I just pushed some more edits, trying to compress some of the loops to make the new IndexSet. Sometimes I'm on the fence about whether explicit for loops are better (you can see exactly what they are doing!) or things like `vcat` are better (shorter code, expressing intent).

**Comment by kshyatt** [comment]
Compact, readable code is the best!

**Comment by emstoudenmire** [comment]
Not sure how readable my `vcat` lines are... maybe if we had slicing for IndexSets it would be a lot more readable.

I just pushed another change to make them more readable I think.

Also found some unused variables! I guess Julia doesn't warn about these things? Would be good to have a code analyzer for Julia.

**Pull request merged**